### DISCUSSION TITLE
Prettier removes Promise<void> type and other annotations in TypeScript arrow functions

### DISCUSSION THREAD

**Discussion by danielbjornadal** [prettier/prettier-vscode#3821] [General]
### Summary

Prettier removes Promise<void> type annotations in TypeScript arrow functions, even with "arrowParens": "always" configured, making it impossible to satisfy TypeScript's requirement for generic type parameters on Promises.

### Steps To Reproduce:

1. Create a new TypeScript file with an interface containing Promise<void> return types
2. Configure Prettier with "arrowParens": "always"
3. Save the file and observe Prettier's formatting
4. Note that TypeScript shows an error: "Generic type 'Promise<T>' requires 1 type argument(s)"

### Expected result

Prettier should preserve the Promise<void> type annotation when formatting TypeScript files, especially when "arrowParens": "always" is set.

Example:
```typescript
interface Example {
  myFunction: (param: string) => Promise<void>;
}
```
### Actual result

Prettier removes the <void> type parameter, resulting in invalid TypeScript:
```typescript
interface Example {
  myFunction: (param: string) => Promise;
}
```

This causes TypeScript to error with "Generic type 'Promise<T>' requires 1 type argument(s)".

### Additional information

The only current workarounds are `// @ts-ignore` or `// prettier-ignore`

Neither solution is ideal for maintaining clean, type-safe code.

**VS Code Version**: 1.95.3

**Prettier Extension Version**: 11.0.0

**OS and version**: Ubuntu 24.04

## Prettier Log Output
```
[INFO] Prettier Options:
{
  "plugins": ["@trivago/prettier-plugin-sort-imports"],
  "importOrder": [...],
  "importOrderSeparation": true,
  "importOrderSortSpecifiers": true,
  "printWidth": 120,
  "arrowParens": "always"
}
[INFO] Prettier Resolution: Prettier Path: /Users/[username]/project/node_modules/prettier
[INFO] Formatting file:///Users/[username]/project/src/types.ts
[INFO] Formatting completed in 0.1s
```

**Comment by madaxen86** [comment]
Same issue here.
Temporary workaround is to remove the plugins. No matter what plugins I think, cause I had 
    "@trivago/prettier-plugin-sort-imports",
    "prettier-plugin-tailwindcss",
The error occurred no matter which I disabled, as long as one is active.

**Comment by IanMitchell** [comment]
I also have noticed this. I have a repo with a reproduction here: https://github.com/ianmitchell/tailwind-example

If you open the repro and navigate [to the page.tsx file](https://github.com/IanMitchell/tailwind-example/blob/main/apps/website/app/page.tsx#L1) and save, the extension will change the `Promise<number>` function return signature to `Promise`. This doesn't happen when prettier is invoked via the CLI.

If I can do anything else to help debug this, happy to help!

**Comment by jjram19** [comment]
+1 here

**Comment by davidy3k** [comment]
+1 here

**Comment by pawel-halat** [comment]
I meet in the same issue after packages.json libraries update. Restarting Prettier VScode extension worked for me.

**Comment by madaxen86** [comment]
I had to remove this line of my vscode `settings.json`
`"prettier.prettierPath": "/usr/local/lib/node_modules/prettier/index.cjs", `
Now it works again.

**Comment by solastley** [comment]
+1 here

Only way to fix was to remove this plugin:
```
"plugins": ["@trivago/prettier-plugin-sort-imports"],
```

which we'd love to keep using...

**Comment by sd-hoag** [comment]
> +1 here
> 
> Only way to fix was to remove this plugin:
> 
> ```
> "plugins": ["@trivago/prettier-plugin-sort-imports"],
> ```
> 
> which we'd love to keep using...

Exact same situation here. Just added the plugin to a new repo and observed the behavior

**Comment by sd-hoag** [comment]
> > +1 here
> > Only way to fix was to remove this plugin:
> > ```
> > "plugins": ["@trivago/prettier-plugin-sort-imports"],
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which we'd love to keep using...
> 
> Exact same situation here. Just added the plugin to a new repo and observed the behavior

Reinstalling the VSCode Prettier extentions _appears_ to have solved this for me

**Comment by vinnymac** [comment]
For me Prettier 3.3.2 or lower does not exhibit this behavior. I narrowed it down to 3.3.3+ that does, which makes me wonder [what changed in 3.3.3](https://github.com/prettier/prettier/compare/3.3.2...3.3.3) that could have broken compatibility when using plugins and generics at the same time.

Uninstalling and reinstalling the extension did not help, but disabling plugins also prevents the generics from being removed, so long as 0 plugins are configured.

Related discussion on SO [here](https://stackoverflow.com/questions/78757481/prettier-is-removing-typescript-generic-annotation-from-react-class-component)

**Comment by cedeber** [comment]
I don't have this issue in vscode, but only in the CLI. But same for me, removing the only plugin I had (prettier-plugin-tailwindcss) was solving the issue in the cli. For information though, weirdly I was used to use pnpm instead of npm and had no issue with pnpm. Not sure why it has to be different but maybe it helps.

Update: I went back to prettier 3.4.2 and it solved the issue. Then I installed 3.5.0 again and it works. Looks like a npm install issue?

**Comment by mgoulet101** [comment]
FWIW, I ran into this same issue after I updated my project typescript dependency from v4.9 to v5.7 and found my way here. The recommendation from @pawel-halat worked for me. I restarted my extensions and the problem went away. I was just about to give up and disable with a prettier-ignore comment.

@cedeber , it might have worked for you because you effectively restarted the extension by switching versions.

**Comment by quark-zju** [comment]
FWIW, `prettier-plugin-sort-imports` seems to have this problem. I switched to `prettier-plugin-organize-imports` and it works fine.

**Comment by quentez** [comment]
I'm having the same problem with `prettier-plugin-organize-imports`. Disabling the plugin solves it.

**Comment by nonpop** [comment]
There was no prettier in my package.json and I had the same problem. A simple `pnpm add -D prettier` fixed it for me (it installed version 3.5.1).

**Comment by SPodjasek** [comment]
@nonpop 😄 Actually, for me, it was the exact opposite. I had the `prettier` package included in my `package.json` and after upgrading a bunch of packages (including prettier) it started to strip my `Promise` types. I didn't notice it at first as my default setting is to format on save, but when I touched a file with `Promise<void>` set as the return type of a function it started to throw compilation errors. I just removed `prettier` from `package.json` as direct dependency and everything went back to normal.
My project uses shared prettier configuration from npm package, and prettier property is set in `package.json` so VSCode extension properly picks it up and uses with built-in prettier extension version.

**Comment by fruitbang** [comment]
I tried to fix it for 6 hours, edited the prettier.config.js, didn't sleep until 4am...
Just had to restart vscode ಥ‿ಥ
all for this >>> "plugins": ["@trivago/prettier-plugin-sort-imports"]

**Comment by vickyzheng2003** [comment]
Upgrade prettier from 3.4.2 to 3.5.2 solved my problem. I use prettier-plugin-organize-imports as well. I use visual studio 2022 and typescript Analyzer(ESlint). Copy new installed prettier to C:\Users\xx\AppData\Local\Microsoft\VisualStudio\17.0_aa37ebb0\Extensions\nwyzzcs5.qrm\Node\node_modules.

**Comment by vinnymac** [comment]
I can also confirm that upgrading beyond 3.3.2 is now possible, I went straight to 3.5.3 without any issues. Curious which release fixed it, but not going to spend the time to find out.

**Comment by Bruno86889** [comment]
> There was no prettier in my package.json and I had the same problem. A simple `pnpm add -D prettier` fixed it for me (it installed version 3.5.1).


it worked for me, i updated the prettier version in my global dependencies with `npm i -g prettier@latest` and restarted VSCode

**Comment by lidingcai** [comment]
I must answer here, same problem. My solution is: **Format Document With**   change     **Configure Default Formatter**   from  **Prettier-Code Format** to **Prettier Eslint**, and the problem solved. I don't know why, but at least, the problem solved

**Comment by sfrog** [comment]
for me, reinstall prettier solve the problem

**Comment by Bespalov-D-A** [comment]
> I meet in the same issue after packages.json libraries update. Restarting Prettier VScode extension worked for me.

It helped me. Thank you!

**Comment by SirDank** [comment]
> There was no prettier in my package.json and I had the same problem. A simple `pnpm add -D prettier` fixed it for me (it installed version 3.5.1).

thanks!

**Comment by trusktr** [comment]
This seems to be flaky. It happened to me today. Then I 

- commented out plugins in prettierrc,
- saved a file in vs code to run prettier and formatting worked,
- uncommented the plugins to re-enable them,
- made and change and saved again to run prettier again

The problem was gone after re-enabling plugins. It was randomly converting `Promise<foo>` into `Promise`, and changing and restoring my prettierrc somehow fixed it in VS Code.

**Comment by abielzulio** [comment]
adding `parser: 'babel-ts` to my prettier config solves this for me (prettier v3.6.2)

https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/338#issuecomment-2574259286

source: https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/338
